### PR TITLE
Remove node/element group fields

### LINKED ISSUE #263: Simplify group fields and subdomains
For the next major release, propose to simplify (and in the process break) group API by removing element and node groups as independent field types so there is only one FieldGroup type.

Benefits:
1. Reduces number of fields, so client doesn't need to see a long list including group.mesh2d etc. which they don't understand.
2. Simplifies coordination of related sub-object groups e.g. adding/removing faces, lines and nodes.
3. Means all groups can be serialised in EX files; currently subobject groups can be created without an owner group, and these are not written to EX.
4. Simplify group API. FieldGroup will have API to directly get/create/destroy MeshGroup and NodesetGroup objects, cutting out the need to get/create FieldElementGroup/FieldNodeGroup then calling getMeshGroup/getNodesetGroup. MeshGroup and NodesetGroup can also have getFieldGroup() method which is lacking the current API (and needed to pass to conditional functions).
5. Simplifies client use; graphics subgroup Field should just be the group; currently confusing that it could be the group or the specific sub-element group for the in-use domain, or even the sub-element group for the wrong domain.
6. Eliminates issue of renaming sub-object group fields when owner group is renamed, which has never been completed.

Example code:

```
group = fieldmodule.createFieldGroup()
meshGroup = group.createMeshGroup(mesh2d)
meshGroup.addElement(face1)
```

Issues:
1. This is a full break. To keep the most common code working (albeit deprecated), we may maintain FieldElementGroup/FieldNodeGroup objects which are not Fields, but which have getMeshGroup/getNodesetGroup methods which work as now. Base field class API for evaluating real values could be offered in the C++/python APIs. [Aside: Field.evaluateBool() should probably be added to the Field class to support group and logical operators going forward.]
2. Reading serialised json description files will require ongoing migration for names of non-existent sub-element groups to that of the group itself. When reading Fieldmodule serialisation, sub-element groups can be ignored. When reading graphics serialisation, field names ending in ``.mesh#d``, ``.nodes`` and ``.datapoints`` can remove the suffix and find the group with the prefix name.

### PR DESCRIPTION
**Pull request by rchristie**
Closes #263.
Remove deprecated API:
Elementtemplate defineFieldSimpleNodal, defineFieldConstant, setting nodes before in template.

### CODE REVIEW

File: src/finite_element/finite_element_conversion.cpp [hunk 1]
```diff
@@ -473,25 +480,18 @@ int Convert_finite_elements_data::convertSubelement(cmzn_element_id element,
...
-									this->addNode(values, node);
+									display_message(ERROR_MESSAGE, "convert elements.  Failed to create node.");
+									return 0;
 								}
-								else
+								if (node)
 								{
-									display_message(ERROR_MESSAGE, "convert elements.  Failed to create node.");
-									return_code = 0;
+									this->addNode(values, node);
```
Comment (hsorby) [inline src/finite_element/finite_element_conversion.cpp]: Surely this should be 
```
if (node)
{
    this->addNode(values, node);
}
else
{
    display_message(ERROR_MESSAGE, "convert elements. Failed to create node.");
    return 0;
}
```

File: tests/fieldmodule/fieldassignment.cpp [hunk 2]
```diff
@@ -107,9 +107,9 @@ TEST(ZincFieldassignment, cubeOffsetScale)
 
     EXPECT_EQ(RESULT_OK, zinc.root_region.readFile(resourcePath("fieldmodule/cube.exformat").c_str()));
 
-	FieldNodeGroup nodeGroup = zinc.fm.createFieldNodeGroup(nodes);
+	FieldGroup nodeGroup = zinc.fm.createFieldGroup();
```
Comment (hsorby) [inline tests/fieldmodule/fieldassignment.cpp]: Probably a misleading variable name here, 'nodeGroup'  is a bit of a hangover from previous API.

**Review by hsorby — COMMENTED** [review]
There are possibly a few things that I find to be inconsistent:
 * Use of the de-reference operator when declaring pointers, should it be next to the type or the variable?
 * Use of macros to deaccess, access objects.  I think I came across one case where a macro was used
 * Use of `cmzn_` in internal API.  The original use for this construct was to avoid conflicts with similar API from another library. If the function is destined to be public one day maybe it should be commented as such.
 
A couple of very minor findings.

If we were to change the value of CMZN_OK, I think we should not be using `return 0` anywhere.

**Comment by rchristie** [comment]
> Use of the de-reference operator when declaring pointers, should it be next to the type or the variable?

I've always thought the full type of the variable should be together, then the name of it, e.g.
```
cmzn_field*& fieldRef
```
So fieldRef is a reference to a field pointer.
I've recently found my new Visual Studio formats this way by default and I want to go with it.
Admittedly, C/C++ associates it with the variable, such as when declaring several variables at once:
```
cmzn_field *bob, **fred;
```
However, I no longer bulk declare variables like this -- it's typical of the C code I've been removing.

File: src/finite_element/finite_element_conversion.cpp (hunk 1 above)
Comment (rchristie) [reply src/finite_element/finite_element_conversion.cpp]: Fixed.

File: tests/fieldmodule/fieldassignment.cpp (hunk 2 above)
Comment (rchristie) [reply tests/fieldmodule/fieldassignment.cpp]: Fixed.

**Review by hsorby — APPROVED** [review]

**Pull request merged**